test/e2e/major_upgrade_v3.go

```diff
@@ -97,12 +99,50 @@ func MajorUpgradeToV3(logger *log.Logger) error {
 				testnet.NoError("failed to get header", err)
 				if resp.Header.Version.App == v3.Version {
 					upgradeComplete = true
+					if upgradedHeight == 0 {
+						upgradedHeight = resp.Header.Height
+					}
 				}
 				logger.Printf("height %v", resp.Header.Height)
 				lastHeight = resp.Header.Height
 			}
 		}
 	}
 
+	// check if the timeouts are set correctly
+	rpcNode := testNet.Nodes()[0]
+	client, err := rpcNode.Client()
+	testnet.NoError("failed to get client", err)
+	blockTimes := make([]time.Duration, 0, 7)
+	var prevBlockTime time.Time
+	for h := upgradeHeightV3 - 4; h <= upgradeHeightV3+4; h++ {
+		resp, err := client.Header(ctx, &h)
+		testnet.NoError("failed to get header", err)
+		blockTime := resp.Header.Time
+		if prevBlockTime.IsZero() {
+			prevBlockTime = blockTime
+			continue
+		}
+
+		blockDur := blockTime.Sub(prevBlockTime)
+		prevBlockTime = blockTime
+		blockTimes = append(blockTimes, blockDur)
+	}
+
+	if len(blockTimes) < 7 {
+		testnet.NoError("", fmt.Errorf("not enough block times collected: %v", len(blockTimes)))
+	}
+
+	startDur := blockTimes[0]
+	endDur := blockTimes[len(blockTimes)-1]
+
+	if startDur < time.Second*10 {
```

Comment:
afaik, the `upgradeHeightV3` is not the exact height at which upgrade to v3 happens, rather A height at which we saw version v3 in the block header (the transition could have happened earlier as well), this means  block at height `upgradeHeightV3 - 4` is not necessarily a v2 block. If we want to know the exact height at which transition has happened, we need to iterate over all the headers, and find that height.

Reply:
I notice that the ticker is set to 10s, which is half the block time of a v2 block. Therefore, it is likely that upgradeHeightV3 - 4 has a v2 app version in its header. However, to ensure clarity and make the test more readable, I suggest explicitly checking the app version in the loop where block times are read.

Reply:
:+1: 

updated in

https://github.com/celestiaorg/celestia-app/pull/3882/commits/8c019af12a85885553a57016969713aa70a47c5e